### PR TITLE
Added commands API

### PR DESCRIPTION
**Pull request by ArjixWasTaken**

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge 
Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge Merge

File: src/api/Commands.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+const getBuiltInCommands = () => (Vencord.Api.Commands as any)._BuiltIn as Command[];
+// const getCategories = () => (Vencord.Api.Commands as any)._BUILTIN_CATEGORIES as Category[];
+
+const getUniqueCommandId = () => {
+    const existingIds = getBuiltInCommands().map(i => parseInt((i as any).id as string));
+    return ((Math.abs(existingIds.sort((a, b) => a - b)[0]) + 1) * -1).toString()
+}
+
+function modifyOpt(opt: Option) {
+    (opt as any).displayName = opt.name;
```
Comment (Vendicated) [inline src/api/Commands.ts:12]: i would put all these on the interface and just make them optional so you aren't required to specify them

Comment (ArjixWasTaken) [reply src/api/Commands.ts:12]: name is internal anyways, only displayName is show to the user, so them being the same is ok
but oh well

File: src/api/Commands.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import type { Command, /*Category,*/ Option } from "./Commands.d"
+
+const getBuiltInCommands = () => (Vencord.Api.Commands as any)._BuiltIn as Command[];
+// const getCategories = () => (Vencord.Api.Commands as any)._BUILTIN_CATEGORIES as Category[];
+
+const getUniqueCommandId = () => {
+    const existingIds = getBuiltInCommands().map(i => parseInt((i as any).id as string));
+    return ((Math.abs(existingIds.sort((a, b) => a - b)[0]) + 1) * -1).toString()
```
Comment (Vendicated) [inline src/api/Commands.ts:8]: i have no idea what this is supposed to do, why not use Discord's Snowflake module to generate an id? findByProps("fromTimestamp").fromTimestamp(Date.now()) or smth like that

File: src/plugins/commandsAPI.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            find: `"giphy","tenor"`,
+            replacement: [
+                // {
+                //     match: /(\w)=(\(.*?\.BUILT_IN)/,
+                //     replace: (m, k, rest) => `${k}=Vencord.Api.Commands._BUILTIN_CATEGORIES=${rest}`
+                // },
+                {
+                    match: /(?<=\w=)([A-Za-z])(\.filter\(\(function\(\w\){return\[(?:(?:"\w+?")|,)+\]\.includes)/,
+                    replace: (m, v, rest) =>
+                        `((()=>Vencord.Api.Commands._BuiltIn=${v})())${rest}`,
```
Comment (Vendicated) [inline src/plugins/commandsAPI.ts:19]: just webpack search it so we don't need this plugin

**Review by Vendicated — CHANGES_REQUESTED** [review]

File: src/api/Commands.ts (hunk 2 above)
Comment (ArjixWasTaken) [reply src/api/Commands.ts:8]: uhh, they are not snowflakes...they are numbers starting from -1 and going down

File: src/plugins/commandsAPI.ts (hunk 3 above)
Comment (ArjixWasTaken) [reply src/plugins/commandsAPI.ts:19]: :pain:

Comment (ArjixWasTaken) [reply src/plugins/commandsAPI.ts:19]: not exposed

Comment (Vendicated) [reply src/plugins/commandsAPI.ts:15]: I see

File: src/plugins/commandsAPI.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+export default definePlugin({
+    name: "CommandsAPI",
+    authors: [Devs.Arjix],
+    description: "Api required by anything that adds commands.",
+    patches: [
+        {
+            find: `"giphy","tenor"`,
+            replacement: [
+                {
+                    match: /(?<=\w=)([A-Za-z])(\.filter\(\(function\(\w\){return\[(?:(?:"\w+?")|,)+\]\.includes)/,
```
Comment (Vendicated) [inline src/plugins/commandsAPI.ts:13]: why use A-Z over \w?

Comment (ArjixWasTaken) [reply src/plugins/commandsAPI.ts:13]: paranoia, \w is fine

File: src/api/Commands.d.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+export type CommandReturnValue = {
+    content: string;
+};
+
+export type Parameter = {
+    type: number;
+    name: string;
+    value: string;
```
Comment (Vendicated) [inline src/api/Commands.d.ts:144]: are you sure this is always a string?

File: src/api/Commands.d.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    USER = 2,
+    MESSAGE = 3,
+}
+
+export type CommandReturnValue = {
+    content: string;
+};
+
+export type Parameter = {
+    type: number;
```
Comment (Vendicated) [inline src/api/Commands.d.ts:142]: shouldnt this be ApplicationCommandOptionType?

File: src/plugins/vencordCommands.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    description: "Built-in vencord commands",
+    authors: [Devs.Arjix],
+    dependencies: ["CommandsAPI"],
+    helloWorld: {
+        name: "hello world",
+        description: "Sends 'Hello world!'",
+        type: ApplicationCommandType.CHAT_INPUT,
+        inputType: ApplicationCommandInputType.BUILT_IN_TEXT,
+        execute: () => ({ content: "Hello world!" }),
+    } as Command,
```
Comment (Vendicated) [inline src/plugins/vencordCommands.ts:21]: looks good but should remove this plugin / change to more useful commands

File: src/api/Commands.d.ts (hunk 6 above)
Comment (ArjixWasTaken) [reply src/api/Commands.d.ts:142]: honestly, its been so long that I have forgotten
+ discord obfuscated everything so i cant easily find the enums again

File: src/api/Commands.d.ts (hunk 5 above)
Comment (ArjixWasTaken) [reply src/api/Commands.d.ts:144]: value? yeah, it is always a string, the type is only used for validation

File: src/api/Commands.d.ts (hunk 6 above)
Comment (Vendicated) [reply src/api/Commands.d.ts:142]: you have the interface in the same file xd just change it to that

oh also please change it from type => interface

Comment (ArjixWasTaken) [reply src/api/Commands.d.ts:142]: > you have the interface in the same file xd just change it to that

you might have misunderstood me, by `so i cant easily find the enums again` i meant that i cant verify that it is correct/incorrect

File: src/plugins/vencordCommands.ts (hunk 7 above)
Comment (ArjixWasTaken) [reply src/plugins/vencordCommands.ts:21]: hmm, are meme commands ok with you?

Comment (ArjixWasTaken) [reply src/plugins/vencordCommands.ts:21]: will go with a lenny face command

Comment (Ruthenic) [reply src/plugins/vencordCommands.ts:21]: uwuifier where?

Comment (ArjixWasTaken) [reply src/plugins/vencordCommands.ts:21]: > uwuifier where?

PR it later, lol

Comment (Ruthenic) [reply src/plugins/vencordCommands.ts:21]: i will as soon as ven merges the fucking pr :imp:

File: src/api/Commands.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        opt.displayDescription = opt.description;
+    opt.options?.forEach(modifyOpt);
+}
+
+export const commands = new Map<string, Command>();
+export function registerCommand(command: Command) {
+    (command as any).id = getUniqueCommandId();
+    (command as any).applicationId = "-1"; // -1 is for the built-in category
+
+    if (command.displayName == undefined) command.displayName = command.name;
```
Comment (Vendicated) [inline src/api/Commands.ts:28]:
```suggestion
    command.displayName ||= command.name;
```
same below

File: src/api/Commands.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    if (command.displayDescription == undefined)
+        command.displayDescription = command.description;
+
+    command.options?.forEach(modifyOpt);
+    commands[(command as any).id + "_" + command.name] = command;
+    getBuiltInCommands().push(command);
+}
+
+export function unregisterCommand(command: Command) {
```
Comment (Vendicated) [inline src/api/Commands.ts:37]: this should only take the command name

Comment (ArjixWasTaken) [reply src/api/Commands.ts:37]: but there can be multiple commands with the same name...

**Review by Vendicated — APPROVED** [review]

**Comment by Vendicated** [comment]
good to merge?

**Pull request merged**